src/java/org/apache/cassandra/transport/Dispatcher.java

```diff
@@ -72,10 +72,33 @@ public void dispatch(Channel channel, Message.Request request, FlushItemConverte
         Stage.NATIVE_TRANSPORT_REQUESTS.submit(() -> processRequest(channel, request, forFlusher, backpressure));
     }
 
-    /**
-     * Note: this method may be executed on the netty event loop, during initial protocol negotiation
-     */
-    static Message.Response processRequest(ServerConnection connection, Message.Request request, Overload backpressure)
+    void processRequest(Channel channel, Message.Request request, FlushItemConverter forFlusher, Overload backpressure)
+    {
+        assert request.connection() instanceof ServerConnection;
+        ServerConnection connection = (ServerConnection) request.connection();
+        Thread requestThread = Thread.currentThread();
+        this.processRequest(connection, request, backpressure)
+            .whenCompleteAsync((response, ex) -> {
+                FlushItem<?> toFlush;
+                if (ex == null)
+                {
+                    toFlush = forFlusher.toFlushItem(channel, request, response);
+                    Message.logger.trace("Responding: {}, v={}", response, connection.getVersion());
+                }
+                else
+                {
+                    JVMStabilityInspector.inspectThrowable(ex);
+                    ExceptionHandlers.UnexpectedChannelExceptionHandler handler = new ExceptionHandlers.UnexpectedChannelExceptionHandler(channel, true);
+                    ErrorMessage error = ErrorMessage.fromException(ex, handler);
+                    error.setStreamId(request.getStreamId());
+                    toFlush = forFlusher.toFlushItem(channel, request, error);
+                }
+                ClientWarn.instance.resetWarnings();
```

Comment:
the client warning is executor local.
Does resetting it work correctly if we change the executor?

Reply:
Actually, why do we need to get back to `Stage.NATIVE_TRANSPORT_REQUESTS.executor` and cannot respond on e.g. READ stage?
Are there any thread safety considerations? I don't think so...

Reply:
> the client warning is executor local.

Good point. The stage executors are of type `LocalAwareExecutorService`, which should propagate the executor locals (I will try to add a test specifically for this). That said, client warnings should probably be reset right after the `processRequest` invocation otherwise (if async) the next native transport task would find them non-null. WDYT?

> Actually, why do we need to get back to Stage.NATIVE_TRANSPORT_REQUESTS.executor and cannot respond on e.g. READ stage?

I moved the response back to the native transport stage because I don't want the "processing" stages to deal with any network-related stuff which could be blocking. To be clear though, this wouldn't change the observation above.